File: internal/client/subscriptions/subscriptions.go
package subscriptions

import (
	"encoding/json"
	"fmt"
	"os"
	"path"
	"path/filepath"
	"strings"

	"github.com/bmatcuk/doublestar/v4"
	"gopkg.in/yaml.v3"
)

const (
	FileName = "syft.sub.yaml"
)

type Action string

const (
	ActionAllow Action = "allow"
	ActionPause Action = "pause"
	ActionBlock Action = "block"
)

type Defaults struct {
	Action Action `yaml:"action" json:"action"`
}

type Rule struct {
	Action   Action `yaml:"action" json:"action"`
	Datasite string `yaml:"datasite,omitempty" json:"datasite,omitempty"`
	Path     string `yaml:"path" json:"path"`
}

type Config struct {
	Version  int      `yaml:"version" json:"version"`
	Defaults Defaults `yaml:"defaults" json:"defaults"`
	Rules    []Rule   `yaml:"rules" json:"rules"`
}

func DefaultConfig() *Config {
	return &Config{
		Version: 1,
		Defaults: Defaults{
			Action: ActionBlock,
		},
		Rules: []Rule{},
	}
}

func (a *Action) UnmarshalYAML(value *yaml.Node) error {
	if value == nil {
		*a = ActionBlock
		return nil
	}
	action, err := parseAction(value.Value)
	if err != nil {
		return err
	}
	*a = action
	return nil
}

func (a *Action) UnmarshalJSON(data []byte) error {
	var raw string
	if err := json.Unmarshal(data, &raw); err != nil {
		return err
	}
	action, err := parseAction(raw)
	if err != nil {
		return err
	}
	*a = action
	return nil
}

func parseAction(raw string) (Action, error) {
	switch strings.ToLower(strings.TrimSpace(raw)) {
	case string(ActionAllow):
		return ActionAllow, nil
	case string(ActionPause):
		return ActionPause, nil
	case string(ActionBlock), "deny":
		return ActionBlock, nil
	case "":
		return ActionBlock, nil
	default:
		return "", fmt.Errorf("invalid action %q", raw)
	}
}

func LoadAction(raw string) (Action, error) {
	return parseAction(raw)
}

func Load(path string) (*Config, error) {
	raw, err := os.ReadFile(path)
	if err != nil {
		if os.IsNotExist(err) {
			return DefaultConfig(), nil
		}
		return nil, err
	}

	var cfg Config
	if err := yaml.Unmarshal(raw, &cfg); err != nil {
		return nil, err
	}
	if cfg.Version == 0 {
		cfg.Version = 1
	}
	if cfg.Defaults.Action == "" {
		cfg.Defaults.Action = ActionBlock
	}
	return &cfg, nil
}

func Save(path string, cfg *Config) error {
	if cfg == nil {
		return fmt.Errorf("config is nil")
	}
	if cfg.Version == 0 {
		cfg.Version = 1
	}
	if cfg.Defaults.Action == "" {
		cfg.Defaults.Action = ActionBlock
	}

	raw, err := yaml.Marshal(cfg)
	if err != nil {
		return err
	}

	if err := os.MkdirAll(filepath.Dir(path), 0o700); err != nil {
		return err
	}

	tmpPath := path + ".tmp"
	if err := os.WriteFile(tmpPath, raw, 0o600); err != nil {
		return err
	}
	return os.Rename(tmpPath, path)
}

func IsSubFile(p string) bool {
	p = normalizePath(p)
	return path.Base(p) == FileName
}

func (c *Config) ActionForPath(owner, relPath string) Action {
	if c == nil {
		return ActionBlock
	}

	relPath = normalizePath(relPath)
	if relPath == "" {
		return c.Defaults.Action
	}

	datasite, rest := splitDatasite(relPath)
	if datasite == "" {
		return c.Defaults.Action
	}
	if isOwnerDatasite(owner, datasite) {
		return ActionAllow
	}

	action := c.Defaults.Action
	fullPath := relPath

	for _, rule := range c.Rules {
		if !rule.matches(datasite, fullPath, rest) {
			continue
		}
		if rule.Action != "" {
			action = rule.Action
		}
	}

	return action
}

func (r Rule) matches(datasite, fullPath, pathWithin string) bool {
	if r.Path == "" {
		return false
	}
	if r.Datasite != "" {
		if !matchGlob(r.Datasite, datasite) {
			return false
		}
		return matchGlob(r.Path, pathWithin)
	}
	return matchGlob(r.Path, fullPath)
}

func matchGlob(pattern, target string) bool {
	pattern = normalizePath(pattern)
	target = normalizePath(target)
	ok, err := doublestar.Match(pattern, target)
	if err != nil {
		return false
	}
	return ok
}

func normalizePath(p string) string {
	if p == "" {
		return ""
	}
	p = strings.ReplaceAll(p, "\\", "/")
	p = strings.TrimLeft(p, "/")
	p = path.Clean(p)
	p = strings.TrimPrefix(p, "./")
	return p
}

func splitDatasite(relPath string) (string, string) {
	parts := strings.SplitN(relPath, "/", 2)
	if len(parts) == 0 {
		return "", ""
	}
	if len(parts) == 1 {
		return parts[0], ""
	}
	return parts[0], parts[1]
}

func isOwnerDatasite(owner, datasite string) bool {
	owner = strings.ToLower(strings.TrimSpace(owner))
	if owner == "" || datasite == "" {
		return false
	}
	return strings.EqualFold(owner, datasite)
}

File: rust/src/subscriptions.rs
use std::path::{Path, PathBuf};

use globset::Glob;
use serde::{Deserialize, Serialize};

const DEFAULT_VERSION: u32 = 1;

#[derive(Clone, Debug, Serialize, Deserialize)]
pub struct Subscriptions {
    #[serde(default = "default_version")]
    pub version: u32,
    #[serde(default)]
    pub defaults: Defaults,
    #[serde(default)]
    pub rules: Vec<Rule>,
}

#[derive(Clone, Debug, Serialize, Deserialize)]
pub struct Defaults {
    #[serde(default)]
    pub action: Action,
}

impl Default for Defaults {
    fn default() -> Self {
        Defaults {
            action: Action::Block,
        }
    }
}

#[derive(Clone, Debug, Serialize, Deserialize)]
pub struct Rule {
    #[serde(default)]
    pub action: Action,
    #[serde(default)]
    pub datasite: Option<String>,
    pub path: String,
}

#[derive(Clone, Debug, Serialize, Deserialize, PartialEq, Eq)]
#[serde(rename_all = "lowercase")]
pub enum Action {
    Allow,
    Pause,
    Block,
    #[serde(other)]
    Unknown,
}

impl Action {
    fn normalize(self) -> Action {
        match self {
            Action::Unknown => Action::Block,
            other => other,
        }
    }
}

impl Default for Action {
    fn default() -> Self {
        Action::Block
    }
}

fn default_version() -> u32 {
    DEFAULT_VERSION
}

pub fn default_config() -> Subscriptions {
    Subscriptions {
        version: DEFAULT_VERSION,
        defaults: Defaults {
            action: Action::Block,
        },
        rules: Vec::new(),
    }
}

pub fn load(path: &Path) -> anyhow::Result<Subscriptions> {
    let raw = match std::fs::read_to_string(path) {
        Ok(raw) => raw,
        Err(err) if err.kind() == std::io::ErrorKind::NotFound => {
            return Ok(default_config())
        }
        Err(err) => return Err(err.into()),
    };
    let mut cfg: Subscriptions = serde_yaml::from_str(&raw)?;
    if cfg.version == 0 {
        cfg.version = DEFAULT_VERSION;
    }
    cfg.defaults.action = cfg.defaults.action.clone().normalize();
    for rule in &mut cfg.rules {
        rule.action = rule.action.clone().normalize();
    }
    Ok(cfg)
}

pub fn save(path: &Path, cfg: &Subscriptions) -> anyhow::Result<()> {
    let cfg = normalize_config(cfg);
    if let Some(parent) = path.parent() {
        std::fs::create_dir_all(parent)?;
    }
    let raw = serde_yaml::to_string(&cfg)?;
    let tmp = path.with_extension("tmp");
    std::fs::write(&tmp, raw)?;
    std::fs::rename(tmp, path)?;
    Ok(())
}

pub fn config_path(data_dir: &Path) -> PathBuf {
    data_dir.join(".data").join("syft.sub.yaml")
}

pub fn is_sub_file(path: &str) -> bool {
    path.trim_end_matches('/')
        .ends_with("/syft.sub.yaml")
        || path == "syft.sub.yaml"
}

pub fn action_for_path(cfg: &Subscriptions, owner: &str, rel_path: &str) -> Action {
    let rel = normalize_path(rel_path);
    let (datasite, rest) = split_datasite(&rel);
    if datasite.is_empty() {
        return cfg.defaults.action.clone().normalize();
    }
    if datasite.eq_ignore_ascii_case(owner) {
        return Action::Allow;
    }

    let mut action = cfg.defaults.action.clone().normalize();
    let full_path = rel.clone();
    for rule in &cfg.rules {
        if !rule_matches(rule, &datasite, &full_path, &rest) {
            continue;
        }
        action = rule.action.clone().normalize();
    }
    action
}

fn rule_matches(rule: &Rule, datasite: &str, full_path: &str, rest: &str) -> bool {
    if rule.path.trim().is_empty() {
        return false;
    }
    if let Some(ds) = &rule.datasite {
        if !matches_glob(ds, datasite) {
            return false;
        }
        return matches_glob(&rule.path, rest);
    }
    matches_glob(&rule.path, full_path)
}

fn matches_glob(pattern: &str, target: &str) -> bool {
    let pattern = normalize_path(pattern);
    let target = normalize_path(target);
    let matcher = match Glob::new(&pattern) {
        Ok(glob) => glob.compile_matcher(),
        Err(_) => return false,
    };
    matcher.is_match(target)
}

fn normalize_path(raw: &str) -> String {
    let mut out = raw.replace('\\', "/");
    while out.starts_with('/') {
        out.remove(0);
    }
    while out.contains("//") {
        out = out.replace("//", "/");
    }
    out
}

fn split_datasite(rel: &str) -> (String, String) {
    if rel.is_empty() {
        return ("".to_string(), "".to_string());
    }
    let mut parts = rel.splitn(2, '/');
    let ds = parts.next().unwrap_or("").to_string();
    let rest = parts.next().unwrap_or("").to_string();
    (ds, rest)
}

fn normalize_config(cfg: &Subscriptions) -> Subscriptions {
    let mut out = cfg.clone();
    out.defaults.action = out.defaults.action.clone().normalize();
    for rule in &mut out.rules {
        rule.action = rule.action.clone().normalize();
    }
    if out.version == 0 {
        out.version = DEFAULT_VERSION;
    }
    out
}
